crates/bevy_render/src/camera/visible_entities.rs

```diff
@@ -24,22 +24,200 @@ impl VisibleEntities {
     }
 }
 
+/// A mask describes which rendering group an entity belongs to.
+///
+/// Cameras with this component will only render entities with a matching
+/// mask.
+///
+/// There are 32 groups numbered `0` - `31`. A mask may belong to one or more
+/// groups, or no group at all.
+///
+/// An entity with a mask belonging to no groups is invisible.
+///
+/// The [`Default`] instance of `RenderingMask` returns a mask belonging to
+/// group `0`, the first group.
+#[derive(Copy, Clone, Reflect, PartialEq, Eq, PartialOrd, Ord)]
+#[reflect(Component)]
+pub struct RenderingMask(u32);
+
+impl std::fmt::Debug for RenderingMask {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("RenderingMask")
+            .field(&self.get_groups())
+            .finish()
+    }
+}
+
+impl std::iter::FromIterator<u8> for RenderingMask {
+    fn from_iter<T: IntoIterator<Item = u8>>(i: T) -> Self {
+        i.into_iter()
+            .fold(RenderingMask(0), |mask, g| mask.with_group(g))
+    }
+}
+
+/// Defaults to a mask belonging to group `0`, the first group.
+impl Default for RenderingMask {
+    fn default() -> Self {
+        RenderingMask(1)
+    }
+}
+
+impl RenderingMask {
+    /// Create a new `RenderingMask` belonging to the given rendering group.
+    pub fn group(n: u8) -> Self {
+        RenderingMask(0).with_group(n)
+    }
+
+    /// Create a new `RenderingMask` that belongs to all rendering groups.
+    pub fn all_groups() -> Self {
+        RenderingMask(u32::MAX)
+    }
+
+    /// Create a new `RenderingMask` that belongs to no rendering groups.
+    pub fn no_groups() -> Self {
+        RenderingMask(0)
+    }
+
+    /// Create a `RenderingMask` from a list of groups.
+    pub fn from_groups(groups: &[u8]) -> Self {
+        groups
+            .iter()
+            .fold(RenderingMask(0), |mask, g| mask.with_group(*g))
+    }
+
+    /// Add the given group to the mask.
+    ///
+    /// This may be called multiple times to allow an entity to belong
+    /// to multiple rendering groups. The maximum group is 31.
+    ///
+    /// # Panics
+    /// Panics when called with a group greater than 31.
+    pub fn with_group(mut self, group: u8) -> Self {
+        assert!(group < 32, "RenderingMask only supports groups 0 to 31");
+        self.0 |= 1 << group;
+        self
+    }
+
+    /// Removes the given rendering group from the mask.
+    ///
+    /// # Panics
+    /// Panics when called with a group greater than 31.
+    pub fn without_group(mut self, group: u8) -> Self {
+        assert!(group < 32, "RenderingMask only supports groups 0 to 31");
+        self.0 |= 0 << group;
+        self
+    }
+
+    /// Get a vector of this mask's groups.
+    pub fn get_groups(&self) -> Vec<u8> {
```

Comment:
Can we make this return a `impl Iterator<Item=u8>` instead of `Vec<u8>`? This would cut down on unnecessary allocations and people needing a Vec can call .collect() on it.

Reply:
And with the "layers" rename we could just rename this method to "iter()"